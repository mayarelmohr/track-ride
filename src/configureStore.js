import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import { persistStore, persistReducer } from "redux-persist";
import localForage from "localforage";
import rootReducer from "./reducers";
import { OrderedMap } from "immutable";
import { createTransform } from "redux-persist";

const SetTransform = createTransform(
  (inboundState) => {
    return { ...inboundState, stations: inboundState.stations.toArray() };
  },
  (outboundState) => {
    return {
      ...outboundState,
      stations: new OrderedMap(outboundState.stations),
    };
  },
  { whitelist: ["trip"] }
);

/* eslint-disable no-underscore-dangle */
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
/* eslint-enable */
const persistConfig = {
  key: "0.1.1",
  storage: localForage,
  whitelist: ["trip"],
  transforms: [SetTransform],
};
const persistedReducer = persistReducer(persistConfig, rootReducer);

const configureStore = (initialState = {}) => {
  const store = createStore(
    persistedReducer,
    initialState,
    composeEnhancers(applyMiddleware(thunk))
  );
  const persistor = persistStore(store);
  return { store, persistor };
};

export const { store, persistor } = configureStore(window.REDUX_STATE || {});
