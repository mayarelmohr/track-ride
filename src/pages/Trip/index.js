import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { csv } from "d3";
import {
  setStations,
  setTripStartTime,
  setTripState,
  updateBookings,
  updateCurrentStation,
  setBookings,
  setTripTime,
} from "../../reducers/trip";
import Loading from "../../components/Common/Loading";
import {
  mapDirectionsToPathSelector,
  getDistanceSelector,
  getStartPointSelector,
} from "../../selectors/stations";
import { TRIP_TIME } from "../../helpers/constants";
import Trip from "../../components/TripWrapper";

const TripWrapper = (props) => {
  const [isLoading, setLoading] = useState(true);
  const initializeTrip = async () => {
    const stationsData = await csv("./data/route.csv");
    props.setStationsAction(stationsData);
    const usersData = await csv("./data/users.csv");
    props.setBookingsAction(usersData);
    props.setTripTimeAction(TRIP_TIME);
  };
  const { isDataReady } = props;

  useEffect(() => {
    if (!isDataReady) {
      /** Cached prop in stations reducer to avoid reinitializing data */
      initializeTrip().then(() => {
        setLoading(false);
      });
    } else {
      setLoading(false);
    }
  }, []);

  if (isLoading) {
    return <Loading />;
  }
  return <Trip />;
};

const mapStateToProps = (state) => {
  const {
    tripTime,
    tripState,
    isDataReady,
    startTime,
    currentStation,
  } = state.trip;
  const paths = mapDirectionsToPathSelector(state);
  const distance = getDistanceSelector(state);
  const bookings =
    currentStation?.bookings || getStartPointSelector(state)?.bookings || [];

  return {
    startTime,
    isDataReady,
    distance,
    tripTime,
    tripState,
    paths,
    bookingsList: bookings,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setStationsAction: (data) => dispatch(setStations(data)),
    setBookingsAction: (data) => dispatch(setBookings(data)),
    updateBookingsAction: (stationId) => dispatch(updateBookings(stationId)),
    setTripTimeAction: (time) => dispatch(setTripTime(time)),
    setTripStateAction: (state) => dispatch(setTripState(state)),
    setTripStartTimeAction: (time) => dispatch(setTripStartTime(time)),
    updateCurrentStationAction: (stationId) =>
      dispatch(updateCurrentStation(stationId)),
  };
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(TripWrapper)
);
