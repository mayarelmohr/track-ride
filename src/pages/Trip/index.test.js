import React from "react";
import { unmountComponentAtNode } from "react-dom";
import { MemoryRouter } from "react-router-dom";
import Trip from "./index";
import { Provider } from "react-redux";
import { render, fireEvent } from "@testing-library/react";
import { store } from "../../configureStore";

const TripElement = (
  <Provider store={store}>
    <MemoryRouter>
      <Trip />
    </MemoryRouter>
  </Provider>
);

test("Shows loading indicator", () => {
  const { container } = render(TripElement);
  expect(container.innerHTML).toMatch("Loading");
});

test("add snapshot for trip loading", () => {
  const { container, asFragment } = render(TripElement);
  expect(asFragment(container)).toMatchSnapshot();
});
