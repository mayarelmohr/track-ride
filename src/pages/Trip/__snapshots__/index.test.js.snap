// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add snapshot for trip loading 1`] = `
<DocumentFragment>
  <p>
    <span
      aria-labelledby="loading"
      role="img"
    >
      ‚è≥
    </span>
    Loading
  </p>
</DocumentFragment>
`;
